src/Shared/BinarySerializer.cs

```diff
@@ -16,10 +19,14 @@ public byte[] Serialize(object data)
             {
                 data = new RedisNull();
             }
-            var binaryFormatter = new BinaryFormatter();
+
             using (var memoryStream = new MemoryStream())
             {
-                binaryFormatter.Serialize(memoryStream, data);
+                // prepend type information to serialized data
+                Type type = data.GetType();
+                var id = System.Text.ASCIIEncoding.ASCII.GetBytes(type.AssemblyQualifiedName + '|');
```

Comment:
nit: I don't know whether AQN can include non-ASCII characters; I'm guessing it *can* - so this should almost certainly be UTF8, not ASCII

Reply:
what's the throughput here? there are a *number* of things we could do to reduce allocations here - avoiding the extra `byte[]`, the `RedisNull` instance, etc. If it doesn't matter: fine

Reply:
I have not attempted to optimize performance yet. I hope to validate the initial approach first.

Reply:
It looks like the approach is valid, so let's optimize its perf before merging.

Reply:
I removed a fair amount of code, so the performance should be okay. I haven't run any tests of the throughput, but I can run some if desired.

Reply:
@mgravell can you recommend any tricks for prepending the header more efficiently? I think this will be a somewhat hot path, so it's worth optimizing as much as possible. But I'm not sure how we could avoid the byte[] allocation